### PR TITLE
Allow fragments in unordered groups

### PR DESCRIPTION
**Pull request by tfonda-fbk**
To fix issue #3068 as hinted by @szarnekow 
Please let me know if I should add a test case (where and what should I test? I'm not familiar with Xtext tests at all).

### CODE REVIEW

**Comment by szarnekow** [comment]
Could you please add tests to https://github.com/eclipse/xtext/blob/main/org.eclipse.xtext.tests/src/org/eclipse/xtext/parser/unorderedGroups/GrammarParserTest.java ?

**Comment by szarnekow** [comment]
And here: https://github.com/eclipse/xtext/blob/main/org.eclipse.xtext.tests/src/org/eclipse/xtext/xtext/XtextValidationTest.java#L1345 (ff)

**Comment by tfonda-fbk** [comment]
@szarnekow 
I've added a test to XtextValidationTest.java (not sure I did it correctly, but it does pass and it does indeed fail if I remove from XtextValidator.java the two lines I'm adding to fix the bug).
I'm unsure how I should structure the test in GrammarParserTest.java, though. I've tried something like this:
```java
	@Test public void testFragments() throws Exception {
		String grammar =
			"grammar test with org.eclipse.xtext.common.Terminals\n" +
			"generate test 'http://test'\n" +
			"MyRule: IntFeature & 'keyword';\n" +
			"fragment IntFeature: myFeature=INT;\n";
		XtextResource resource = getResourceFromString(grammar);
		Assert.assertTrue(resource.getErrors().isEmpty());
	}
```
but the parsing of this grammar fails for some reason, even though I copy-pasted the concept from another test class:
```
java.lang.AssertionError: XtextSyntaxDiagnostic: null:3 no viable alternative at input 'grammar'
XtextSyntaxDiagnostic: null:3 missing ';' at 'test'
XtextSyntaxDiagnostic: null:3 mismatched input 'with' expecting ':' expected:<0> but was:<3>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.eclipse.xtext.testing.AbstractXtextTests.getResourceAndExpect(AbstractXtextTests.java:289)
	at org.eclipse.xtext.testing.AbstractXtextTests.getResource(AbstractXtextTests.java:275)
	at org.eclipse.xtext.testing.AbstractXtextTests.getResource(AbstractXtextTests.java:246)
	at org.eclipse.xtext.testing.AbstractXtextTests.getResourceFromString(AbstractXtextTests.java:238)
	at org.eclipse.xtext.parser.unorderedGroups.GrammarParserTest.testFragments(GrammarParserTest.java:136)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:530)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:758)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:453)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:211)
```
I think I have to override several methods to make this work, the same way it is done in the class I copied the concept from. But maybe there is a simpler way? Any advice?
Thanks.

**Comment by cdietrich** [comment]
the grammar that is parsd is

grammar org.some.Grammar with org.eclipse.xtext.common.Terminals
generate name 'http://my.uri'
RootRule: grammar test with org.eclipse.xtext.common.Terminals
generate test 'http://test'
MyRule: IntFeature & 'keyword';
fragment IntFeature: myFeature=INT;
;

see org.eclipse.xtext.parser.unorderedGroups.GrammarParserTest.surroundWithGrammar(String)

**Comment by tfonda-fbk** [comment]
@cdietrich 
Thanks! I've adjusted the string:
```java
	@Test public void testFragments() throws Exception {
		String grammar =
			"MyRule;\n" +
			"MyRule: IntFeature & 'keyword';\n" +
			"fragment IntFeature: myFeature=INT\n";
		XtextResource resource = getResourceFromString(grammar);
		Assert.assertTrue(resource.getErrors().isEmpty());
	}
```
but now the test always passes, no matter if my fix is present or not. I've even inserted a new validator that always and unconditionally raises an error, and it still passes. It looks like the XtextValidator is completely bypassed. Any idea what's going on?

**Comment by cdietrich** [comment]
if the test does not call the validator, it just tests the grammar part

**Comment by cdietrich** [comment]
@HannesWell @LorenzoBettini do you have an idea why
Build / Test Results (pull_request) 
is marked as failed?

**Comment by tfonda-fbk** [comment]
@cdietrich I've implemented the second test calling the validator by hand very naively. The test now works as expected. Please let me know if it looks good to you.

File: org.eclipse.xtext.tests/src/org/eclipse/xtext/parser/unorderedGroups/GrammarParserTest.java [hunk 1]
```diff
@@ -126,6 +131,21 @@ public void setUp() throws Exception {
...
+	@Test public void testFragments() throws Exception {
+		String grammar =
+			"MyRule;\n" +
+			"MyRule: IntFeature & 'keyword';\n" +
+			"fragment IntFeature: myFeature=INT\n";
+		XtextResource resource = getResourceFromString(grammar);
+		Iterable<EObject> contents = () -> resource.getAllContents();
+		XtextValidator validator = get(XtextValidator.class);
+		boolean result = true;
+		for (final EObject eObject : contents) {
```
Comment (cdietrich) [inline org.eclipse.xtext.tests/src/org/eclipse/xtext/parser/unorderedGroups/GrammarParserTest.java:143]: am, not sure about that @szarnekow 
should validation be done with XtextValidationTest?

Comment (szarnekow) [reply org.eclipse.xtext.tests/src/org/eclipse/xtext/parser/unorderedGroups/GrammarParserTest.java:143]: Sorry, I was on the phone when I provided the code pointers. I might have missed that the validator wasn’t used from one of them. I think adding the parser test is still ok even though it doesn’t fail without the fix. It ensures that it won’t fail in the future, too.

**Comment by tfonda-fbk** [comment]
@szarnekow @cdietrich
Thanks, I have restored the previous, always-passing version of the test.

**Review by szarnekow — APPROVED** [review]
Thank you for the initiative!

**Pull request merged**